galley/pkg/config/source/kube/apiserver/tombstone/recover.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tombstone
+
+import (
+	"fmt"
+	"reflect"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/pkg/log"
+
+	"istio.io/istio/galley/pkg/config/source/kube/apiserver/stats"
+)
+
+var scope = log.RegisterScope("source", "", 0)
+
+// RecoverResource from a kubernetes tombstone (cache.DeletedFinalStateUnknown). Returns the resource or nil if
+// recovery failed.
+func RecoverResource(obj interface{}) metav1.Object {
```

Comment:
This is a pretty generic function and I believe could be shared with other apps that talk to k8s (e.g. Pilot) ... maybe we move this to a top-level kube utility package?

Reply:
I thought we wanted Pilot to be agnostic eventually? I'd prefer not to split this code just yet, unless there is an explicit need. I'd much rather move all of galley/pkg/config to be a top-level package.

Reply:
sure that's fine.